### ISSUE TITLE
bug(ipeps): U(1)-Sz simple update collapses the state to exactly zero at every D (survival time grows with D); dense is unaffected

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#865]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Found while building a control measurement for #853. Not a #853 defect and not caused by #851 — filed separately so both stay scoped.

## The state becomes exactly zero

D=3, U(1)-Sz, `heisenberg_u1sz_init_pair(D=3, key=PRNGKey(0))`, `dt=0.05`, the four-phase checkerboard sweep:

```
|A| = |B| = 0.0     and     lam_h_AB = lam_h_BA = lam_v_AB = lam_v_BA = [0, 0, 0]
```

Not small — **exactly** zero. Everything downstream then behaves as you would expect from a zero tensor: `ctm_tensor` returns a zero corner, the raw RDM has trace 0, and normalising it gives `NaN`.

## Which branch, and what changes

Same fixture, same four-phase loop driven by hand against each tree's `_simple_update_2site_*` so the comparison is like-for-like:

| tree | tensors die at | `min over lambdas of max(lam)` |
|---|---|---|
| `main` (sqrt(sigma) baked into Gamma) | **never** — `\|A\| = 1.0` throughout | hits **0.0 at step 24** and stays there |
| #844 (bare Gamma, two lambdas) | **step 22** | 1.0 → 2.73e-04 (step 20) → 1.53e-85 (step 22) |
| #851 (bare Gamma, four lambdas) | **step 42** | 1.0 → 4.33e-07 (step 40) → 1.21e-51 (step 42) |

Read across the table:

* **A bond's Schmidt spectrum collapses to all-zero around step 22-24 on every tree.** That is the underlying problem and it predates both PRs — it is #667's product-state collapse in its U(1)-Sz form.
* **On `main` that is survivable.** `sqrt(sigma)` is scaled into `Gamma`, so the site tensor keeps its own weight: you get a degenerate state with one dead bond, which is bad but is still a state.
* **#844 makes it fatal.** Its (correct, and separately motivated) change to keep `Gamma` *bare* moves all the weight into the lambdas, so a lambda of zero multiplies the whole tensor to zero on the next absorb. A degenerate answer becomes no answer.
* **#851 roughly doubles the survival time** (22 → 42) by giving the four bonds their own spectra instead of stamping one onto both, but it does not cure it.

So this is a **#844 regression on the symmetric path**, on top of a pre-existing defect. #844's own #667 guard is a `DenseTensor` test, which is why nothing caught it.

## Likely amplifier

`lam_new = sigma / (jnp.max(sigma) + EPS)` with `EPS = 1e-15`, in both `_simple_update_2site_horizontal_tensor` and its vertical twin.

That additive absolute epsilon is a no-op while `max(sigma) >> EPS`, and inverts the moment it is not: at `max(sigma) ~ EPS` the "normalised" lambda comes out at `max(sigma)/EPS < 1` instead of 1, the next sweep absorbs that shrunken lambda into `theta`, `sigma` shrinks again, and it runs away. The measured trajectories are exactly two steps wide — 2.7e-04 → 1.5e-85, and 4.3e-07 → 1.2e-51 — which is the signature of a runaway rather than a decay.

This is the same defect class as #748 (additive absolute epsilon in a denominator). A relative floor, or a `jnp.where(smax > 0, smax, 1.0)`, would keep the normalisation exact and leave `sigma = 0` returning zeros rather than corrupting the healthy case.

**But the epsilon is only the amplifier.** It does not explain why `max(sigma)` reaches 1e-15 at all, given that `A` and `B` are renormalised to unit norm every step and every lambda is max-normalised. That part is unexplained and is the actual fix.

## Possibly the same bug

`examples/su_symmetric_ctm_e2e.py` documents a "PART B (currently BROKEN)" in which a U(1)-Sz SU state fed to `ctm_tensor_2site` gives a corner of norm 0 and an energy of exactly 0, and attributes it to "a general 2-site forward-CTM defect". It runs D=4 for 60 steps — well past the step-22 death above. If the SU output is already exactly zero, the CTM is not the defect there and that note is misattributed. Worth checking before anyone spends time on the 2-site forward CTM.

## Reproduction

```python
jax.config.update("jax_enable_x64", True)
A, B = heisenberg_u1sz_init_pair(D=3, key=jax.random.PRNGKey(0))
gate = _make_trotter_gate_tensor(heisenberg_gate_u1sz(), 0.05, site_tensor=A)
A, B, lam = _simple_update_checkerboard_sweep(A, B, gate, 3, 200)   # or the hand-rolled 4-phase loop
assert float(A.norm()) == 0.0
```

Refs #844, #851, #667, #748, #853.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## This is not a D=3 bug: the U(1)-Sz sweep collapses at **every** D, and dense collapses at none

Measured on the #863 four-bond sweep (`dt=0.05`, 160 phases, death = any bond's
`max(lambda)` reaching 0 or `|A| = 0`), scanning D=2..8 in both representations
against the same driver. The baseline reproduces this issue's trajectory
bit-exactly first (`4.3276e-07 -> 1.2082e-51` at D=3), and a transcription
control pins the phase loop to `_simple_update_checkerboard_sweep` at zero
difference, so the numbers below come from this issue's own driver.

| D | parity | bal | virtual charges | **symmetric** | **dense** (final min lam) |
|---|---|---|---|---|---|
| 2 | even | no | `[0, 1]` | dies **phase 10** | survives (1.59e-01) |
| 3 | odd | yes | `[0, 1, -1]` | dies **phase 43** | survives (8.85e-02) |
| 4 | even | yes | `[0, 1, -1, 0]` | dies **phase 25** | survives (3.90e-01) |
| 5 | odd | no | `[0, 1, -1, 0, 1]` | dies **phase 66** | survives (3.31e-01) |
| 6 | even | yes | `[0, 1, -1, 0, 1, -1]` | survives\* (2.33e-04) | survives (5.83e-01) |
| 7 | odd | yes | `[0, 1, -1, 0, 1, -1, 0]` | dies **phase 152** | survives (6.91e-01) |
| 8 | even | no | `[0, 1, -1, 0, 1, -1, 0, 1]` | survives\* (3.91e-03) | survives (7.61e-01) |

\* survives the 160-phase budget only — `min lam` is already 1e-4..1e-3 and
falling, i.e. on the same trajectory, not stable.

Two things fall out:

**D=3 is not special, and it is not even the worst case.** D=2 dies at phase 10,
**four times sooner** than the D=3 this issue documents. D=3 was measured because
that is what was in front of us.

**Lifetime grows with D, in both parities** — odd 43 -> 66 -> 152, even 10 -> 25
-> 160+. That is the trap: at D=6 or D=8 with a typical step budget the sweep
looks fine, which is plausibly why this went unnoticed. It is not fine; the
spectrum is on the same decay, just slower.

**Dense simple update is healthy at every D**, and its worst Schmidt value
*improves* with D (1.59e-01, 8.85e-02, 3.90e-01, 3.31e-01, 5.83e-01, 6.91e-01,
7.61e-01) — the exact opposite of the symmetric arm. So this is not a defect in
simple update's dynamics; it is specific to the U(1)-Sz path.

### "odd D" is not a coherent class here, for anyone tempted by the pattern

`heisenberg_u1sz_init_pair` cycles `[0, 1, -1][i % 3]`, so the bond structure
repeats **mod 3, not mod 2**. D=3 gets a balanced `[0,1,-1]`; D=5 gets
`[0,1,-1,0,1]` with two `+1` and one `-1`. Charge balance does not predict the
outcome either — unbalanced D=8 outlives balanced D=7.

The structurally suspicious case is **D=2**, whose virtual charges are `[0, 1]`
with no `-1` sector at all, so the bond cannot carry negative Sz flux — and it
is the one that dies fastest. The docstring warns about a *different* trap
("pure ±1 virtual charges produce a parity obstruction") and does not cover
this one.

### Suggested retitle

> `bug(ipeps): U(1)-Sz simple update collapses the state to exactly zero at every D (survival time grows with D); dense is unaffected`

The `#844`-regression analysis in the issue body still stands — bare `Gamma`
turns a dead bond from degenerate into fatal — as does the `EPS = 1e-15`
additive-floor amplifier. What changes is the scope: the underlying collapse is
not indexed by D, so a fix validated only at D=3 will not be validated at all.

### Also relevant

Belief propagation does **not** rescue this, in either form. Against the phase-43
baseline, #870's `bp_gauge_checkerboard` as shipped dies at phase 4; with its
overflow defect fixed (reported separately on that PR) it dies at phase 6, each
within ~2 phases of the first BP call. BP converges cleanly onto a spectrum
whose minimum is already ~1e-27 — re-deriving the weights reproduces the
collapse rather than escaping it, which says the runaway lives in the state and
not in the cached weights.
